M2/Macaulay2/m2/webapp.m2

```diff
@@ -1,38 +1,47 @@
-    webAppTextTag):=      -- other text
-apply((17,18,19,20,28,30),ascii)
-
-
-htmlWithTexLiteral = s -> replace("\\\\","&bsol;",htmlLiteral s);
-
-texWrap := x -> concatenate("\\(",x,"\\)")
-
-htmlWithTex Thing := x -> texWrap("\\displaystyle " | texMath x) -- by default, for KaTeX we use tex (as opposed to html)
+    (webAppEndTag,            -- closing tag ~ </span>
+	webAppHtmlTag,        -- indicates what follows is HTML ~ <span class='M2Html'>
+	webAppOutputTag,      -- it's html but it's output ~ <span class='M2Html M2Output'>
+	webAppInputTag,       -- it's text but it's input ~ <span class='M2Input'>
+	webAppInputContdTag,  -- text, continuation of input
+	webAppTextTag,        -- other text ~ <span class='M2Text'>
+	webAppTexTag,         -- TeX start ~ \(
+	webAppTexEndTag       -- TeX end ~ \)
+	)=apply((17,18,19,20,28,30,31,17),ascii);
+
+texMathStart := "\\(";
+texMathEnd := "\\)";
+texWrap := x -> texMathStart | texMath x | texMathEnd; -- similar to 'tex', but only used by webapp.m2 to avoid thread-safety issues
+
+htmlWithTex Thing := texWrap -- by default, we use tex (as opposed to html)
 
 -- text stuff: we use html instead of tex, much faster (and better spacing)
 htmlWithTex Hypertext := html
--- the % is relative to line-height
-htmlWithTex Net := n -> concatenate("<pre><span style=\"display:inline-table;vertical-align:", 
-    toString(100*(height n-1)), "%\">", apply(unstack n, x-> htmlWithTexLiteral x | "<br/>"), "</span></pre>")
-htmlWithTex String := x -> concatenate("<pre>", htmlWithTexLiteral x, "</pre>") -- only problem is, this ignores starting/ending \n. but then one should use Net for that
-htmlWithTex Descent := x -> concatenate("<span style=\"display:inline-table\"><pre>", sort apply(pairs x,
+htmlWithTex Net := n -> concatenate("<pre style=\"display:inline-table;vertical-align:",
+    toString(100*(height n-1)), "%\">\n", apply(unstack n, x-> htmlLiteral x | "<br/>"), "</pre>") -- the % is relative to line-height
+htmlWithTex String := x -> concatenate("<pre style=\"display:inline\">\n", htmlLiteral x, "</pre>",
+    if #x>0 and last x === "\n" then "<br/>") -- fix for html ignoring trailing \n
+htmlWithTex Descent := x -> concatenate("<pre style=\"display:inline-table\">\n", sort apply(pairs x,
      (k,v) -> (
 	  if #v === 0
-	  then toString k -- sucks but no choice
-	  else toString k | " : " | htmlWithTex v
-	  ) | "<br/>"), "</pre></span>")
--- some expressions can be htmlWithTex'ed directly w/o reference to texMath
-htmlWithTex RowExpression := x -> concatenate("<span>",apply(toList x, htmlWithTex),"</span>")
-htmlWithTex Holder := x -> htmlWithTex x#0
+	  then htmlWithTex net k -- sucks but no choice
+	  else htmlWithTex net k | " : " | htmlWithTex v
+	  ) | "<br/>"), "</pre>")
+
+-- now preparation for output
+
+webAppBegin = (displayStyle) -> (
+    texMathStart = webAppTexTag | (if displayStyle then "\\displaystyle " else "");
+    texMathEnd = webAppTexEndTag;
```

Comment:
What's the point of changing the values of these two variables?

Reply:
it's still convenient to have these variables (which are internal to webapp.m2 now) because Print and AfterPrint have slightly different styles of output. and yes, it makes the output routines themselves not thread-safe, but I really don't see the problem -- certainly output routines should not be called in parallel, it would create a mess anyway.